### PR TITLE
2.0 preview

### PR DESCRIPTION
**Pull request by valdrinkoshi**
Updated to work both in 1.0 and 2.0.

Focus wrapping is already broken in 1.x when an overlay hosts elements with `tabindex > 0` & there isn't a robust solution, see more details [here](https://github.com/PolymerElements/iron-overlay-behavior/pull/241). Updated README.md with possible workarounds.

Tests updated to wait for MutationObservers to be done, as `Polymer.dom.flush()` won't be able to flush them in native ShadowDom anyways (see https://github.com/Polymer/polymer/issues/4036).

~~Tests failing on MSEdge until https://github.com/Polymer/web-component-tester/issues/528 is fixed~~

### CODE REVIEW

File: iron-overlay-manager.html [hunk 1]
```diff
@@ -220,6 +220,7 @@
       }
       this.backdropElement.style.zIndex = this._getZ(overlay) - 1;
       this.backdropElement.opened = !!overlay;
+      this.backdropElement.prepare();
```
Comment (e111077) [inline iron-overlay-manager.html]: Why is this necessary now?

File: test/iron-overlay-behavior.html [hunk 2]
```diff
@@ -27,14 +27,16 @@
     <link rel="import" href="test-buttons.html">
     <link rel="import" href="test-menu-button.html">
 
-    <style is="custom-style">
+    <custom-style><style is="custom-style">
       iron-overlay-backdrop {
         /* For quicker tests */
         --iron-overlay-backdrop: {
           transition: none;
+          /* Safari 10 requires prefix */
```
Comment (e111077) [inline test/iron-overlay-behavior.html]: image

File: test/iron-overlay-behavior.html [hunk 3]
```diff
@@ -27,14 +27,16 @@
     <link rel="import" href="test-buttons.html">
     <link rel="import" href="test-menu-button.html">
 
-    <style is="custom-style">
+    <custom-style><style is="custom-style">
```
Comment (e111077) [inline test/iron-overlay-behavior.html]: style nit

File: test/iron-overlay-behavior.html [hunk 4]
```diff
@@ -1110,25 +1115,37 @@
...
-                var btn = Polymer.dom(overlay2).querySelector('button');
-                btn.addEventListener(eventName, overlay2.close.bind(overlay2));
-                MockInteractions.tap(btn);
-                assert.isFalse(overlay2.opened, 'overlay2 closed');
-                assert.isTrue(overlay1.opened, 'overlay1 opened');
-                overlay2.addEventListener('iron-overlay-closed', function() {
-                  assert.isTrue(overlay1.opened, 'overlay1 still opened');
-                  done();
-                });
+        test('click event handled only by top overlay', function(done) {
```
Comment (e111077) [inline test/iron-overlay-behavior.html]: 👍 👍 👍 👍 💯

**Comment by e111077** [comment]
Also, tests are not passing. What's up with that?

**Comment by valdrinkoshi** [comment]
> Tests failing on MSEdge because of Polymer/polyserve#161

File: iron-focusables-helper.html [hunk 5]
```diff
@@ -127,18 +127,17 @@
         //  </div>
         // TODO(valdrin) support ShadowDOM v1 when upgrading to Polymer v2.0.
```
Comment (valdrinkoshi) [inline iron-focusables-helper.html]: Not really. So this helper does give you a sorted list of tabbable nodes; that example above this TODO explains what would happen with native shadow dom. I should definitely get back to this & fix it.

File: iron-focusables-helper.html [hunk 6]
```diff
@@ -127,18 +127,17 @@
...
         var children;
-        if (element.localName === 'content') {
+        if (element.localName === 'content' || element.localName === 'slot') {
```
Comment (valdrinkoshi) [inline iron-focusables-helper.html]: We cannot, as this method will search deeply in the shadow roots of the elements; in Polymer 1.x there might be elements that are still using `<content>` - e.g. some element not updated yet to hybrid (user created Polymer Element)

File: iron-overlay-manager.html (hunk 1 above)
Comment (valdrinkoshi) [inline iron-overlay-manager.html]: Property observers are not triggered until the element is attached https://github.com/Polymer/polymer/issues/4526
I should probably add a comment here..!

File: iron-focusables-helper.html (hunk 6 above)
Comment (e111077) [inline iron-focusables-helper.html]: 👍

Comment (valdrinkoshi) [inline iron-focusables-helper.html]: Note to self:
taborder w/o ShadowDOM: http://output.jsbin.com/rotohan/3/
taborder in SD v0: http://output.jsbin.com/tavedi/3/
taborder in SD v1: http://output.jsbin.com/vaciwaw/3/

**Comment by valdrinkoshi** [comment]
Note to self for updating `IronFocusablesHelper` - check native behavior in Chrome of taborder:
taborder w/o ShadowDOM: http://output.jsbin.com/rotohan/3/
taborder in SD v0: http://output.jsbin.com/tavedi/3/
taborder in SD v1: http://output.jsbin.com/vaciwaw/3/

File: iron-focusables-helper.html (hunk 5 above)
Comment (valdrinkoshi) [inline iron-focusables-helper.html]: So turns out this implementation is buggy also in SD v0 when there are elements with tabindex > 0, so most likely I'll either leave this as is for now. More details https://github.com/PolymerElements/iron-overlay-behavior/pull/241

Comment (e111077) [inline iron-focusables-helper.html]: 👍 understandable

**Comment by valdrinkoshi** [comment]
https://github.com/Polymer/polyserve/issues/161 is fixed & released but `wct` still needs to update to that version. Tracking issue: https://github.com/Polymer/web-component-tester/issues/528

**Comment by e111077** [comment]
This works on everything else but edge; I'll approve since that polyserve fix is taking a long time to propagate to polymer-cli@next
